[AMDGPU] Add intrinsic readanylane

GinShio opened the pull request:
Sometimes, we know the value is uniform, but backend cannot easily prove that it is uniform.

This change introduces the intrinsic `readanylane`, which is similar to readfirstlane, but has a couple of advantages:
 + It can be sunk into control flow.
 + If the result is needed in a vgpr then the v_readfirstlane instruction can be optimized away.

CC: @arsenm, @jayfoad, @nhaehnle, @ruiling

Commit:
[AMDGPU] Add intrinsic readanylane

Sometimes, we know the value is uniform, but backend cannot easily prove
that it is uniform.

This change introduces the intrinsic `readanylane`, which is similar to
readfirstlane, but has a couple of advantages:
 + It doesn't convergent, so can be moved between control flows.
 + If the result is needed in a vgpr then the v_readfirstlane
   instruction can be optimized away.

jayfoad commented:
Thanks for working on this!

> It doesn't convergent, so can be moved between control flows.

Unfortunately the new intrinsic still has to be `convergent`.

Consider these transformations:
```
A: x = def();                            B: x = def();
   y = read*lane(x);                        if (divergentcondition) {
   if (divergentcondition) {                  y = read*lane(x);
     use(y);                                  use(y);
   }                                        }
```
Transforming A-->B is not safe for `readfirstlane`, but it is safe for `readanylane` (because if x is uniform outside the `if` then it must also be uniform inside the `if`). Transforming B-->A is not safe for `readfirstlane` and not safe for `readanylane`.

Unfortunately the `convergent` attribute currently restricts both of these transforms, so we need `readanylane` to be convergent to prevent B-->A.

Maybe in future we can have some better way to express this which will allow transforming A-->B but prevent B-->A.

Cc: @nhaehnle @ssahasra

GinShio commented:
> Transforming A-->B is not safe for `readfirstlane`, but it is safe for `readanylane` (because if x is uniform outside the `if` then it must also be uniform inside the `if`). Transforming B-->A is not safe for `readfirstlane` and not safe for `readanylane`.

Sorry, I don't quite understand why the latter is unsafe. (B-->A for `readanylane`)
In my understand, if x is uniform, `readanylane` is not affected by exec flag.

```mermaid
block-beta
  columns 8
  IfTh0[" "] IfTh1["42"] IfTh2["42"] IfTh3["42"] IfTh4["42"] IfTh5[" "] IfTh6["42"] IfTh7["42"]
  ThenTh0[" "] ThenTh1["42"] ThenTh2["42"] ThenTh3["42"] ThenTh4["42"] ThenTh5[" "] ThenTh6["42"] ThenTh7["42"]
  ElseTh0[" "] ElseTh1["42"] ElseTh2["42"] ElseTh3["42"] ElseTh4["42"] ElseTh5[" "] ElseTh6["42"] ElseTh7["42"]

  classDef disable fill:#808080,stroke:#333;
  classDef convergence fill:#218733,stroke:#333;
  classDef then fill:#A93376,stroke:#333;
  classDef else fill:#65CBEE,stroke:#333;
  class IfTh0,IfTh5,IfTh7 disable
  class IfTh1,IfTh2,IfTh3,IfTh4,IfTh6,IfTh7 convergence
  class ThenTh0,ThenTh1,ThenTh2,ThenTh4,ThenTh5 disable
  class ThenTh3,ThenTh6,ThenTh7 then
  class ElseTh0,ElseTh3,ElseTh5,ElseTh6,ElseTh7 disable
  class ElseTh1,ElseTh2,ElseTh4 else
```

For example, the 1st line is *if-block*, then it is *then-block*, and the last *else-block*.
Looks like reading data is a safe behavior no matter which bb it is in.

Could you help me point out my mistake?

jayfoad commented:
The definition of readanylane(x) has to say something like "_if x is uniform_ then the result is x, otherwise the result is undefined". The first part "if x is uniform" means that it is uniform at the place where it is used, not at the place where it is defined. In your example that means that all the dark red cells must have the same value (42) but other green cells could be different.

(Another example of why we have to consider uniformity _at the point of use_ is temporal divergence. You can have a uniform def inside a loop with a divergent exit condition, but a use of that def outside the loop must be considered as divergent.)

Commit:
[AMDGPU] Add attribute convergent for readanylane

New intrinsic has to be convergent. It\'s uniform uniform at the place
where it is used.

This commit changes:
 + Add attribute `convergent`.
 + Remove testcase `test_readanylane_hoist`. The intrinsic is no longer
   allowed hoist.

GinShio commented:
Thanks a lot! Added attribute `convergent`.

jayfoad commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2152,6 +2152,13 @@ def int_amdgcn_readfirstlane :
+// control flow. The result is undefined if the value is actual divergent.
+def int_amdgcn_readanylane :
+  Intrinsic<[llvm_any_ty], [LLVMMatchType<0>],
+            [IntrConvergent, IntrNoCallback, IntrNoFree, IntrNoMem, IntrSpeculatable,
```
Did you add IntrSpeculatable deliberately? If so, you should add a test showing what sort of transform it enables.

GinShio replied:
It's removed.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -97,9 +98,11 @@ bool AMDGPUInstructionSelector::isVCC(Register Reg,
 
 bool AMDGPUInstructionSelector::constrainCopyLikeIntrin(MachineInstr &MI,
                                                         unsigned NewOpc) const {
+  const bool NeedExec = NewOpc != AMDGPU::SI_READANYLANE;
```
Why doesn't readanylane need exec?

GinShio replied:
For the 1st version, it's SAlu and I thought it's uniform when defined. So...
Now, the intrinsic depends on exec.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -4658,6 +4658,8 @@ AMDGPURegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
       [[fallthrough]];
     }
+    case Intrinsic::amdgcn_readanylane:
+      [[fallthrough]];
```
Don't need `[[fallthrough]]` when the `case` lines are consecutive.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -546,6 +546,10 @@ def SI_MASKED_UNREACHABLE : SPseudoInstSI <(outs), (ins),
 }
 
+def SI_READANYLANE : SPseudoInstSI <(outs SReg_32:$dst), (ins VGPR_32:$src)> {
+  let SALU = 1;
```
Should be a VALU instruction like readfirstlane is. Maybe define it with `VPseudoInstSI`?

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1081,6 +1081,7 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
   }
   case Intrinsic::amdgcn_permlane64:
   case Intrinsic::amdgcn_readfirstlane:
+  case Intrinsic::amdgcn_readanylane:
```
Add some tests for this in `test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll`

ruiling commented:
> The first part "if x is uniform" means that it is uniform at the place where it is used, not at the place where it is defined. In your example that means that all the dark red cells must have the same value (42) but other green cells could be different.

I would say let's say it clearly that readanylane is defined only if the input is uniform when defined and at the use of readanylane call. We really want the intrinsic to be not convergent to allow related middle-end optimizations.

btw, I think for the existing uniformity analysis, we can say that: "if a value is uniform at use, then it should be uniform at define", right? We only have the case that a value is defined uniform but divergent at its use currently.

ssahasra commented:
> I would say let's say it clearly that readanylane is defined only if the input is uniform when defined and at the use of readanylane call. We really want the intrinsic to be not convergent to allow related middle-end optimizations.

The change description starts with:

> "Sometimes, we know the value is uniform, but backend cannot easily prove that it is uniform."

Can this be solved with an attribute instead?

If we do need an intrinsic, we should stop comparing it with readfirstlane, and just focus on what readanylane means. Would it be correct to rename it to something like "assumeuniform"? Does this provide information that can be propagated backwards when initializing uniformity analysis?

jayfoad commented:
> I would say let's say it clearly that readanylane is defined only if the input is uniform when defined and at the use of readanylane call. We really want the intrinsic to be not convergent to allow related middle-end optimizations.

I'm not comfortable with saying "the input is uniform _when defined_". The problem is with code like this:
```
  x = uniform_def();
  y = divergent_def();
  if (x == y) {
    ...
  }
```
Inside the body of the `if`, generic optimizations should be allowed to replace uses of `x` with `y` and vice versa. But if they replace `readanylane(x)` with `readanylane(y)` then that would break the "uniform when defined" condition.

I don't have a good example of an optimization that actually does this, but I do think they should be allowed to do this.

> btw, I think for the existing uniformity analysis, we can say that: "if a value is uniform at use, then it should be uniform at define", right? We only have the case that a value is defined uniform but divergent at its use currently.

Yes I think that's correct, currently.

> Can this be solved with an attribute instead?

I'm not sure how that would work. We want to apply this to a specific IR `Use`, or to an IR `Value` at a specific point in the CFG. The use case is quite similar to the existing `@llvm.assume` intrinsic.

> If we do need an intrinsic, we should stop comparing it with readfirstlane, and just focus on what readanylane means.

Yeah. I would define it something like "if the argument is uniform, return that value, otherwise the result is undefined" with an explanatory note saying why you would want to use it.

> Would it be correct to rename it to something like "assumeuniform"?

I guess so. Personally I like the name `readanylane`, but that might be because I came up with it :-)

> Does this provide information that can be propagated backwards when initializing uniformity analysis?

Yes, if UA did any backwards propagation then I think this could be a good thing to propagate backwards from.

ssahasra commented:
```
A: x = def();                            B: x = def();
   y = read*lane(x);                        if (divergentcondition) {
   if (divergentcondition) {                  y = read*lane(x);
     use(y);                                  use(y);
   }                                        }
```

I am not sure what is the primary goal here. If the goal is to allow certain defs to be sunk into divergent control flow, then maybe this particular situation is sufficient justification to add a new attribute that allows allow sinking. This can be combined with builtin that declares "uniform at use". A combination of these two will allow the `A --> B` transformation.

In program B, being able to say that a value is uniform at its use inside divergent control-flow can be useful in itself. Trying to hoist this value like in `B --> A` is hard anyway. This should be considered separately from the `A --> B` transform. In the program B, if the user wanted the value to be hoisted, then they probably have sufficient information to write A instead.

Syntax-wise, I am leaning more towards something that looks like `@llvm.assume`. In fact, we could consider `@llvm.is_uniform` as a builtin predicate, and then the user can say `@llvm.assume(@llvm.is_uniform())`. This is similar to the existing `@llvm.is.constant.*` builtins.

jayfoad commented:
> If the goal is to allow certain defs to be sunk into divergent control flow, then maybe this particular situation is sufficient justification to add a new attribute that allows allow sinking. This can be combined with builtin that declares "uniform at use". A combination of these two will allow the A --> B transformation.

Would this new attribute be in addition to `convergent`, or instead of it? I still think `readanylane` should be `convergent` because it depends on the set of active threads, so I see this as a special case where we know the rules can be relaxed: in general `readanylane` should not be moved past divergent control flow, but in the specific case of sinking into an `if` it is OK.

FYI previously, probably before most of your work on convergence, someone proposed splitting `convergent` into two separate attributes, one to prevent sinking (adding additional control dependencies) and one to prevent hoisting (removing control dependencies).

Let me add one more example:
```
A: x = def();                           B: x = def();                           C: x = def();
   y = read*lane(x);                       if (divergentcondition) {               y = read*lane(x);
   if (divergentcondition) {                 y = read*lane(x);                     if (divergentcondition) {
     use(y);                                 use(y);                                 z = read*lane(x);
   }                                       }                                         use(z);
                                                                                   }
```
CSE will transform C-->A if we don't prevent it in some way. That's one of the reasons for marking `readanylane` as `convergent`.

Commit:
[AMDGPU] Remove unused attribute for readanylane

 + Remove unused attribute `speculatable`.
 + Change machine instruction from SAlu to VAlu, add attributes.
 + Add testcase for InstCombine, and CSE.
 + Update docs.

GinShio commented:
Sorry, my understanding of `readanylane` may be gap. But, I prefer `@llvm.assume` if we want to mark the value is uniform when defined.

> Let me add one more example:
> 
> ```
> A: x = def();                           B: x = def();                           C: x = def();
>    y = read*lane(x);                       if (divergentcondition) {               y = read*lane(x);
>    if (divergentcondition) {                 y = read*lane(x);                     if (divergentcondition) {
>      use(y);                                 use(y);                                 z = read*lane(x);
>    }                                       }                                         use(z);
>                                                                                    }
> ```
> 
> CSE will transform C-->A if we don't prevent it in some way. That's one of the reasons for marking `readanylane` as `convergent`.

This example (`C --> A`) maybe good if I understand correctly when `readanylane` used. If we think x is uniform in if-block (y) and also uniform in then-block (z), we can safely do CSE in this case.

I thought we must prevent optimize to GVN hoisting (Following case that `B --> A`, `uniform when used`). It's one of reasons.
```
A: x = def();                            B: x = def();
   y = readanylane(x);                     if (divergentcondition) {
   if (divergentcondition) {                  y = readanylane(x);
     use(y);                                  dosomthing0(y);
   }                                        } else {
                                              z = readanylane(x);
                                              dosomthing1(z);
                                            }
```

Finally, I updated the code. Currently, `readanylane` hits that value is uniform when used.

jayfoad commented:
> > Let me add one more example:
> > ```
> > A: x = def();                           B: x = def();                           C: x = def();
> >    y = read*lane(x);                       if (divergentcondition) {               y = read*lane(x);
> >    if (divergentcondition) {                 y = read*lane(x);                     if (divergentcondition) {
> >      use(y);                                 use(y);                                 z = read*lane(x);
> >    }                                       }                                         use(z);
> >                                                                                    }
> > ```
> > CSE will transform C-->A if we don't prevent it in some way. That's one of the reasons for marking `readanylane` as `convergent`.
> 
> This example (`C --> A`) maybe good if I understand correctly when `readanylane` used. If we think x is uniform in if-block (y) and also uniform in then-block (z), we can safely do CSE in this case.

The problem is when `x` is uniform inside the `if` but divergent outside the `if`. Then `z` is defined but `y` is undefined, so it is not OK to replace a use of `z` with a use of `y`.

Commit:
[AMDGPU] Update testcase for readanylane

Incorrectly added FileCheck pattern for readanylane

jayfoad commented:
> `y = readanylane(x)` that means x is uniform outside of if, and `z = readanylane(x)` that means x is uniform inside of if.
> Looks like okay, OTOW, x is always uniform in this case, either inside or outside of the if. Right?

I don't think it is helpful to say "readanylane(x) _means_ x is uniform", because the presence of the readanylane() call does not magically make x uniform if it was not already uniform. It is better to describe it as "readanylane(x) is x if x is uniform, otherwise readanylane(x) is undefined".

With that in mind, does my explanation make sense to you:

> The problem is when x is uniform inside the if but divergent outside the if. Then z is defined but y is undefined, so it is not OK to replace a use of z with a use of y.

GinShio commented:
> With that in mind, does my explanation make sense to you:

Yes, that's a good explanation.

> because the presence of the readanylane() call does not magically make x uniform if it was not already uniform.

IIUC, if x is not uniform outside of if and we used `readanylane`, the transform `C-->A`:
 + is okay, because it's undefined behavior if src of `readanylane` is non-uniform, we can assume it's uniform when used (magically make x uniform). In this case, user shouldn't the intrinsic if they don't guarantee x is uniform when using `readanylane`.
 + is bad, because it's undefined result, we cannot assume that.

We expect the latter.

ssahasra commented:
> Would this new attribute be in addition to `convergent`, or instead of it? I still think `readanylane` should be `convergent` because it depends on the set of active threads, so I see this as a special case where we know the rules can be relaxed: in general `readanylane` should not be moved past divergent control flow, but in the specific case of sinking into an `if` it is OK.

There should probably be a family of attributes, where `convergent` is the most conservative. The other attributes will add nuance that relaxes what `convergent` means. For example, the property that we need here is that "this convergent call is valid for the subset of threads that implicitly reach here, but not for a superset". This is more specific than saying things like "may sink" or "may add control dependencies". Alternatively we could even implement this as metadata that goes along with the `convergent` attribute.

ssahasra commented:
> It is better to describe it as "readanylane(x) is x if x is uniform, otherwise readanylane(x) is undefined".

Compared to that, `@llvm.assume(@llvm.is.uniform())` actively tells the compiler that the user _knows_ that this use of the value is uniform. But then, if the user is wrong, then the result is undefined anyway. In what way is one better than the other?

GinShio commented:
> > It is better to describe it as "readanylane(x) is x if x is uniform, otherwise readanylane(x) is undefined".
> 
> Compared to that, `@llvm.assume(@llvm.is.uniform())` actively tells the compiler that the user _knows_ that this use of the value is uniform. But then, if the user is wrong, then the result is undefined anyway. In what way is one better than the other?

Sorry, I don't focus on this for long time. Thanks a lot for testing the `llvm.is.uniform`!!!
We discussed offline: "Yes, we know the value is uniform, we wish the value can be hoisted or other." A typical case is `subgroup add`.
We will try to use `is.uniform`. That's cool!

For this inst, maybe it's not what we want, but also we can continue improve it if necessary...

Thanks again to you and jay for detailed explanation.

jayfoad commented:
> > It is better to describe it as "readanylane(x) is x if x is uniform, otherwise readanylane(x) is undefined".
> 
> Compared to that, `@llvm.assume(@llvm.is.uniform())` actively tells the compiler that the user _knows_ that this use of the value is uniform. But then, if the user is wrong, then the result is undefined anyway. In what way is one better than the other?

Are you asking me? I am slightly reluctant to use `@llvm.assume` because I am less familiar with it, and it seems more complicated: to tell whether a value `%x` is uniform you would have to search its uses looking for the `@llvm.assume(@llvm.is.uniform(%x))` pattern, and probably check that those intrinsic calls dominate the use of `%x`. I guess InstCombine is already set up to do this for you, but I am not sure about other parts of the compiler that might want to do an ad hoc optimization when they can assume `%x` is uniform.
